### PR TITLE
Add CI

### PR DESCRIPTION
**Pull request by MichaelStubbings**
Add continuous integration to PandABlocks-server

### CODE REVIEW

File: tests/test_exchange.py [hunk 1]
```diff
@@ -22,32 +22,43 @@
 args = parser.parse_args()
 
 
+print('Defining server socket...')
```
Comment (Araneidae) [inline tests/test_exchange.py:25]: All these prints look like debugging residue left behind.

File: tests/run_with_server [hunk 2]
```diff
@@ -20,6 +20,7 @@ while ! echo '*IDN?' | nc localhost 8888; do
     echo connecting...
     sleep 0.2
 done
+echo 'Connected to server'
```
Comment (Araneidae) [inline tests/run_with_server:23]: Debugging?

File: .github/scripts/config-file-server.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+# Generates config files in PandABlocks-server and server repositories and populates them with information.
+
+PLATFORM=$1
+
+# Determine the toolchain to use
+if [ "$PLATFORM" == "zynq" ]; then
+    TOOLCHAIN=gcc-arm-10.2-2020.11-x86_64-arm-none-linux-gnueabihf
+elif [ "$PLATFORM" == "zynqmp" ]; then
+    TOOLCHAIN=gcc-arm-10.2-2020.11-x86_64-aarch64-none-linux-gnu
+fi
```
Comment (Araneidae) [inline .github/scripts/config-file-server.sh:11]: What if `$PLATFORM` doesn't match?

**Review by Araneidae — CHANGES_REQUESTED** [review]
I have three general comments:

- Can we please squash the numerous CI commits together
- Please remove the debugging lines; I think after this none of the files in the `tests` directory should be included in this merge
- There seem to be a number of significant changes that aren't part of CI.  Hopefully they'll be easier to see after squashing the CI commits.  These changes seem to be:
  - Change of zpkg name.  This affects other parts of the projects, so needs to be a clearly separate commit.
  - Removing the `! nm` line from the `sim_server` target.  Again, hopefully there is a clear commit for this, but I can't see it at the moment!

File: tests/Makefile [hunk 4]
```diff
@@ -5,7 +5,6 @@ TOP = ..
 
 # ------------------------------------------------------------------------------
 # Logic LUT (FPGA LookUp Table) parser test.  This involves generating and
-# running an intermediate executable file.
```
[the remark this answers is not in the document]
Comment (MichaelStubbings) [reply tests/Makefile:8]: Good question, Ill add it back

File: tests/test_exchange.py (hunk 1 above)
Comment (MichaelStubbings) [reply tests/test_exchange.py:25]: Yes, I shall remove them

File: CONFIG.example [hunk 5]
```diff
@@ -7,9 +7,9 @@
 #
 # BUILD_DIR = $(TOP)/build
 
-# Python interpreter.  Default interpreter is python2, needs to be 2.7
+# Python interpreter.  Default interpreter is python3
 #
-# PYTHON = python2
+# PYTHON = python3
```
[the remark this answers is not in the document]
Comment (MichaelStubbings) [reply CONFIG.example:12]: I would suggest updating the default in the Makefile to python3. Do you think that would affect any of the other processes?

[the remark this answers is not in the document]
Comment (MichaelStubbings) [reply CONFIG.example:12]: Or are you referring to the comment above? "Python interpreter. Default interpreter is python3"

File: .github/scripts/config-file-server.sh (hunk 3 above)
Comment (MichaelStubbings) [reply .github/scripts/config-file-server.sh:13]: Added else condition to report mismatch and halt workflow

File: Makefile [hunk 6]
```diff
@@ -151,7 +151,7 @@ clean-docs:
 
 ZPKG_DEPENDS = $(PANDA_KO) $(SERVER) $(SLOW_LOAD) $(DOCS_BUILD_DIR)/index.html
 
-ZPKG = $(BUILD_DIR)/panda-server@$(GIT_VERSION).zpg
+ZPKG = $(BUILD_DIR)/panda-server@$(PLATFORM)-$(GIT_VERSION).zpg
```
[the remark this answers is not in the document]
Comment (MichaelStubbings) [reply Makefile]: Understood, this change will be managed from https://github.com/PandABlocks/PandABlocks-server/pull/18

File: CONFIG.example (hunk 5 above)
[the remark this answers is not in the document]
Comment (MichaelStubbings) [reply CONFIG.example:12]: Discussed in review meeting, Updated python default in Makefile to python3

**Comment by Araneidae** [comment]
@tomtrafford, please don't do documentation commits on this branch.  I'm sorting out the changes here, and I think the safest solution is going to be a manual merge which I'm working on.

**Comment by Araneidae** [comment]
I have manually separated out the various commits in this branch and have squashed the hundred-odd CI commits into one.  This is now merged separately, so I'll delete this branch.

**Pull request closed without merging**